collect_app/src/main/java/org/odk/collect/android/formentry/QuitFormDialogFragment.java

```diff
@@ -0,0 +1,135 @@
+
+    private FormSaveViewModel viewModel;
+    private final ViewModelProvider.Factory viewModelFactory = new FormSaveViewModel.Factory();
+
+    private ListView listView;
+    private String title;
+
+    @Override
+    public void onAttach(@NonNull Context context) {
+        super.onAttach(context);
+        viewModel = ViewModelProviders.of(requireActivity(), viewModelFactory).get(FormSaveViewModel.class);
+
+        title = viewModel.getFormName();
+        if (title == null) {
+            title = context.getString(R.string.no_form_loaded);
+        }
+
+        List<IconMenuItem> items;
+        if ((boolean) AdminSharedPreferences.getInstance().get(AdminKeys.KEY_SAVE_MID)) {
+            items = ImmutableList.of(new IconMenuItem(R.drawable.ic_save, R.string.keep_changes),
+                    new IconMenuItem(R.drawable.ic_delete, R.string.do_not_save));
+        } else {
+            items = ImmutableList.of(new IconMenuItem(R.drawable.ic_delete, R.string.do_not_save));
+        }
+
+        listView = DialogUtils.createActionListView(context);
+
+        final IconMenuListAdapter adapter = new IconMenuListAdapter(context, items);
+        listView.setAdapter(adapter);
+        listView.setOnItemClickListener((parent, view, position, id) -> {
+            IconMenuItem item = (IconMenuItem) adapter.getItem(position);
+
+            if (item.getTextResId() == R.string.keep_changes) {
+                viewModel.saveForm(getActivity().getIntent().getData(), InstancesDaoHelper.isInstanceComplete(false),
+                        null, true);
+
+            } else {
+                ExternalDataManager manager = Collect.getInstance().getExternalDataManager();
+                if (manager != null) {
+                    manager.close();
+                }
+
+                if (viewModel.getAuditEventLogger() != null) {
+                    viewModel.getAuditEventLogger().logEvent(AuditEvent.AuditEventType.FORM_EXIT, true, System.currentTimeMillis());
+                }
+
+                viewModel.removeTempInstance();
+                MediaManager.INSTANCE.revertChanges();
+
+                String action = getActivity().getIntent().getAction();
+                if (Intent.ACTION_PICK.equals(action) || Intent.ACTION_EDIT.equals(action)) {
+                    // caller is waiting on a picked form
+                    Uri uri = InstancesDaoHelper.getLastInstanceUri(viewModel.getAbsoluteInstancePath());
+                    if (uri != null) {
+                        getActivity().setResult(RESULT_OK, new Intent().setData(uri));
+                    }
+                }
+                getActivity().finish();
+            }
+
+            if (getDialog() != null) {
+                getDialog().dismiss();
+            }
+        });
+    }
+
+    @NonNull
+    @Override
+    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
+        super.onCreateDialog(savedInstanceState);
+        setRetainInstance(true);
```

Comment:
You shouldn't need this. Good to look at the docs and have a think about it but it's best to avoid.

Reply:
Thanks @seadowg for directing to the docs. I was thinking of using it and then dismissing the dialog in `onDestroy`, so as to confirm that the dialog is destroyed when the fragment is finally detached, and not on just orientation change. But I found that it is deprecated now, so I think it is better not to use it and rely more on ViewModel.